backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,163 @@
+		vh.refreshFunc[viewName](db)
+	}
+}
+
+// getViewsToRefresh creates a list of views that should be refreshed due to table changes.
+// The viewsToRefresh is the map of tables that shows which views should be refreshed due to table changes.
+func (vh *ViewRefreshHandler) getViewsToRefresh() map[string]bool {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table := range tables {
+		for _, viewName := range vh.viewsToRefresh[table] {
+			viewToRefresh[viewName] = true
+		}
+	}
+
+	return viewToRefresh
+}
+
+func (vh *ViewRefreshHandler) RegisterViewRefreshHandlers(dbDriver string) {
+	switch dbDriver {
+	case DBDriverTypePostgres:
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(applicationViewName),
+			applicationViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(resourceViewName),
+			resourceViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(packageViewName),
+			packageViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(vulnerabilityViewName),
+			vulnerabilityViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+	}
+}
+
+func (vh *ViewRefreshHandler) registerViewRefreshHandler(f refreshFunc, viewName string, tables []string) {
+	vh.refreshFunc[viewName] = f
+	vh.addViewsToRefreshByTable(viewName, tables)
+}
+
+func (vh *ViewRefreshHandler) IsSetViewRefreshHandler() bool {
+	return len(vh.refreshFunc) > 0
+}
+
+func (vh *ViewRefreshHandler) addViewsToRefreshByTable(viewName string, tables []string) {
+	for _, table := range tables {
+		vh.viewsToRefresh[table] = append(vh.viewsToRefresh[table], viewName)
+	}
+}
+
+func (db *Handler) SetMaterializedViewHandler(config *DBConfig) {
+	db.ViewRefreshHandler = &ViewRefreshHandler{
+		tableChanged:              make(map[string]bool),
+		viewsToRefresh:            make(map[string][]string),
+		viewRefreshIntervalSecond: time.Duration(config.ViewRefreshIntervalSecond) * time.Second,
+		refreshFunc:               map[string]refreshFunc{},
+	}
+}
+
+func (db *Handler) RefreshMaterializedViews() {
+	if !db.ViewRefreshHandler.IsSetViewRefreshHandler() {
+		return
+	}
+	for {
+		<-time.After(db.ViewRefreshHandler.viewRefreshIntervalSecond)
+		db.ViewRefreshHandler.runRequiredRefreshes(db.DB)
+	}
+}
+
+func getPostgresViewRefreshHandlerFunc(viewName string) func(db *gorm.DB) {
+	return func(db *gorm.DB) {
+		if err := db.Exec(fmt.Sprintf(refreshMaterializedViewConcurrentlyCommand, viewName)).Error; err != nil {
```

Comment:
When I was talking about concurrency before I wasn't refering to this postgres's concurrency, when I talked about concurrency before I meant:

```
func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
	viewNames := vh.getViewsToRefresh()
	var wg sync.WaitGroup
	for viewName := range viewNames {
	        wg.Add(1)
	        go func() {
	                defer wg.Done()
		        vh.refreshFunc[viewName](db)
		}()
	}
	wg.Wait()
}
```

unsure if gorm/postgres will handle us doing that, but it seems like we should need to serially run the Exec's and then wait on them to be done before calling the next one.

Postgres's CONCURRENTLY option doesn't actually refresh the view concurrently, it just doesn't lock the table during the refresh.

Reply:
Ok, done